File: Day010_2025/Q010_Leetcode.py
from typing import List

class Solution:
    def count(self, s: str) -> List[int]:
        ans = [0] * 26
        for c in s:
            ans[ord(c) - ord('a')] += 1
        return ans

    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        # Q. Word Subsets
        bmax = [0] * 26
        for b in words2:
            bCount = self.count(b)
            for i in range(26):
                bmax[i] = max(bmax[i], bCount[i])

        result = []
        for a in words1:
            aCount = self.count(a)
            if all(aCount[i] >= bmax[i] for i in range(26)):
                result.append(a)

        return result


# Example usage
if __name__ == "__main__":
    words1 = ["amazon", "apple", "facebook", "google", "leetcode"]
    words2 = ["e", "o"]
    sol = Solution()
    print(sol.wordSubsets(words1, words2))

File: Day010_2025/Q010_Leetcode.cpp
class Solution {
public:
    vector<int> countt(string s) {
        vector<int> ans(26, 0);
        for (char c : s)
            ans[c - 'a']++;
        return ans;
    }

    vector<string> wordSubsets(vector<string>& words1, vector<string>& words2) {
        // Q. Word Subsets
        vector<int> bmax(26, 0);
        for (string b : words2) {
            vector<int> bCount = countt(b);
            for (int i = 0; i < 26; ++i)
                bmax[i] = max(bmax[i], bCount[i]);
        }

        vector<string> ans;
        for (string a : words1) {
            vector<int> aCount = countt(a);
            bool isUniversal = true;
            for (int i = 0; i < 26; ++i) {
                if (aCount[i] < bmax[i]) {
                    isUniversal = false;
                    break;
                }
            }
            if (isUniversal)
                ans.push_back(a);
        }

        return ans;
    }
};